contracts/AMMWrapper.sol

```diff
@@ -196,82 +175,36 @@ contract AMMWrapper is IAMMWrapper, ReentrancyGuard, BaseLibEIP712, SignatureVal
-        // used to bypass stack too deep error.
-        txMetaData.subsidyFactor = uint16(subsidyFactor);
-        txMetaData.feeFactor = uint16(_feeFactor);
-        if (!permStorage.isRelayerValid(tx.origin)) {
-            txMetaData.feeFactor = (txMetaData.subsidyFactor > txMetaData.feeFactor) ? txMetaData.subsidyFactor : txMetaData.feeFactor;
-            txMetaData.subsidyFactor = 0;
-        }
-
         // Assign trade vairables
-        internalTxData.fromEth = (order.takerAssetAddr == ZERO_ADDRESS || order.takerAssetAddr == ETH_ADDRESS);
-        internalTxData.toEth = (order.makerAssetAddr == ZERO_ADDRESS || order.makerAssetAddr == ETH_ADDRESS);
-        if (_isCurve(order.makerAddr)) {
+        internalTxData.fromEth = (_order.takerAssetAddr == ZERO_ADDRESS || _order.takerAssetAddr == ETH_ADDRESS);
+        internalTxData.toEth = (_order.makerAssetAddr == ZERO_ADDRESS || _order.makerAssetAddr == ETH_ADDRESS);
+        if (_isCurve(_order.makerAddr)) {
             // PermanetStorage can recognize `ETH_ADDRESS` but not `ZERO_ADDRESS`.
-            // Convert it to `ETH_ADDRESS` as passed in `order.takerAssetAddr` or `order.makerAssetAddr` might be `ZERO_ADDRESS`.
-            internalTxData.takerAssetInternalAddr = internalTxData.fromEth ? ETH_ADDRESS : order.takerAssetAddr;
-            internalTxData.makerAssetInternalAddr = internalTxData.toEth ? ETH_ADDRESS : order.makerAssetAddr;
+            // Convert it to `ETH_ADDRESS` as passed in `_order.takerAssetAddr` or `_order.makerAssetAddr` might be `ZERO_ADDRESS`.
+            internalTxData.takerAssetInternalAddr = internalTxData.fromEth ? ETH_ADDRESS : _order.takerAssetAddr;
+            internalTxData.makerAssetInternalAddr = internalTxData.toEth ? ETH_ADDRESS : _order.makerAssetAddr;
         } else {
-            internalTxData.takerAssetInternalAddr = internalTxData.fromEth ? address(weth) : order.takerAssetAddr;
-            internalTxData.makerAssetInternalAddr = internalTxData.toEth ? address(weth) : order.makerAssetAddr;
+            internalTxData.takerAssetInternalAddr = internalTxData.fromEth ? address(weth) : _order.takerAssetAddr;
+            internalTxData.makerAssetInternalAddr = internalTxData.toEth ? address(weth) : _order.makerAssetAddr;
         }
 
-        txMetaData.transactionHash = _verify(order, _sig);
+        txMetaData.transactionHash = _verify(_order, _sig);
 
-        _prepare(order, internalTxData);
+        _prepare(_order, internalTxData);
 
-        // minAmount = makerAssetAmount * (10000 - subsidyFactor) / 10000
-        uint256 _minAmount = order.makerAssetAmount.mul((BPS_MAX.sub(txMetaData.subsidyFactor))).div(BPS_MAX);
-        (txMetaData.source, txMetaData.receivedAmount) = _swap(order, internalTxData, _minAmount);
+        // amountOutMin = makerAssetAmount * (10000 + feeFactor) / 10000
+        uint256 _amountOutMin = _order.makerAssetAmount.mul((LibConstant.BPS_MAX.add(feeFactor))).div(LibConstant.BPS_MAX);
```

Comment:
Would it be harder to settle when we require an additional fee on swap out amount? 

I think maybe ensuring to swap out at least `makerTokenAmount` is enough. As long as maker knows we charge fee, it's reasonable to give maker a slightly less amount.

Reply:
I was worried about user get the wrong idea when they receive less than the quote. But this can be adjusted in the front end as well. Let's check the current front end policy and align with it.